### PR TITLE
feat: recursion support

### PR DESCRIPTION
**Pull request by polvalente**
Improves the ideas from #61 for handling more edge cases

### CODE REVIEW

**Comment by polvalente** [comment]
@emilsoman there are some minor details to fix, but the module_compiler tests are working correctly!

File: lib/fika/compiler/type_checker.ex [hunk 1]
```diff
@@ -429,6 +510,39 @@ defmodule Fika.Compiler.TypeChecker do
...
 
+  defp reset_scope_and_set_signature(env, {:function, _, {_name, args, _type, _}} = ast) do
+    current_signature = Fika.Compiler.TypeChecker.function_ast_signature(env[:module], ast)
+
+    env
+    |> Map.merge(%{
+      scope: %{},
+      has_effect: false,
+      current_signature: current_signature,
+      latest_called_function: if(env[:first_pass], do: env[:current_signature]),
```
Comment (emilsoman) [inline lib/fika/compiler/type_checker.ex:521]: Do we set `first_pass` to true anywhere?

Comment (polvalente) [reply lib/fika/compiler/type_checker.ex:521]: It was being set to true as the default value for the struct. This might need a refactor now that the struct has been removed.

init_env still sets it, but most tests don't call it
